pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,979 @@
+		// Print ugly but useful diff of expected and received objects for
+		// easier debugging.
+		return fmt.Errorf("Volume check failed [A-expected, B-got]: %s", diff.ObjectDiff(expectedMap, actualMap))
+	}
+	return nil
+}
+
+// checkClaims compares all expectedClaims with set of claims at the end of the
+// test and reports differences.
+func (r *volumeReactor) checkClaims(t *testing.T, expectedClaims []*api.PersistentVolumeClaim) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	expectedMap := make(map[string]*api.PersistentVolumeClaim)
+	actualMap := make(map[string]*api.PersistentVolumeClaim)
+	for _, c := range expectedClaims {
+		c.ResourceVersion = ""
+		expectedMap[c.Name] = c
+	}
+	for _, c := range r.claims {
+		// We must clone the claim because of golang race check - it was
+		// written by the controller without any locks on it.
+		clone, _ := conversion.NewCloner().DeepCopy(c)
+		c = clone.(*api.PersistentVolumeClaim)
+		c.ResourceVersion = ""
+		actualMap[c.Name] = c
+	}
+	if !reflect.DeepEqual(expectedMap, actualMap) {
+		// Print ugly but useful diff of expected and received objects for
+		// easier debugging.
+		return fmt.Errorf("Claim check failed [A-expected, B-actual result]: %s", diff.ObjectDiff(expectedMap, actualMap))
+	}
+	return nil
+}
+
+// checkClaims compares all expectedEvents with events generated during the test
+// and reports differences.
+func checkEvents(t *testing.T, expectedEvents []string, ctrl *PersistentVolumeController) error {
+	var err error
+	fakeRecorder := ctrl.eventRecorder.(*record.FakeRecorder)
+	for i, expected := range expectedEvents {
+		if len(fakeRecorder.Events) <= i {
+			t.Errorf("Event %q not emitted", expected)
+			err = fmt.Errorf("Events do not match")
+			continue
+		}
+		received := fakeRecorder.Events[i]
+		if !strings.HasPrefix(received, expected) {
+			t.Errorf("Unexpected event received, expected %q, got %q", expected, received)
+			err = fmt.Errorf("Events do not match")
+		}
+	}
+	for i := len(expectedEvents); i < len(fakeRecorder.Events); i++ {
+		t.Errorf("Unexpected event received: %q", fakeRecorder.Events[i])
+		err = fmt.Errorf("Events do not match")
+	}
+	return err
+}
+
+// popChange returns one recorded updated object, either *api.PersistentVolume
+// or *api.PersistentVolumeClaim. Returns nil when there are no changes.
+func (r *volumeReactor) popChange() interface{} {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if len(r.changedObjects) == 0 {
+		return nil
+	}
+
+	// For debugging purposes, print the queue
+	for _, obj := range r.changedObjects {
+		switch obj.(type) {
+		case *api.PersistentVolume:
+			vol, _ := obj.(*api.PersistentVolume)
+			glog.V(4).Infof("reactor queue: %s", vol.Name)
+		case *api.PersistentVolumeClaim:
+			claim, _ := obj.(*api.PersistentVolumeClaim)
+			glog.V(4).Infof("reactor queue: %s", claim.Name)
+		}
+	}
+
+	// Pop the first item from the queue and return it
+	obj := r.changedObjects[0]
```

Comment:
yeah, Go's grammar just doesn't allow fields of objects to be part of a := expression.